app/(candidate)/onboarding/[slug]/[step]/components/ballotOffices/BallotRaces.js

```diff
@@ -59,57 +79,63 @@ export default function BallotRaces(props) {
     }
   };
 
+  const clearState = () => {
+    setSelected(false);
+    setInputValue('');
+    setLevel('')
+    setYearFilter(null)
+  }
+
   const handleZipChange = async (newZip) => {
     if (newZip !== zip) {
-      setLoading(true);
-      const newRaces = await fetchRaces(newZip);
-      setRaces(newRaces.races);
-      setFilteredRaces(newRaces.races);
-      setZip(newZip);
-      setSelected(false);
-      setLoading(false);
-      setFilters('');
-      const updatedCampaign = {
+      setZip(newZip)
+      clearState()
+      await loadRaces(newZip)
+      await updateCampaign({
         ...campaign,
         details: {
           ...campaign.details,
           zip: newZip,
         },
-      };
-      await updateCampaign(updatedCampaign);
+      });
     }
   };
 
-  const filterRaces = (value, updatedLevel) => {
-    if (!value && updatedLevel === '') {
-      setFilteredRaces(races);
-      return;
+  const filterRace = (race) => {
+    let positionLevel = race.position.level;
+
+    // TODO: move this kind of data-rewite closer to "source of truth", preferably
+    //  whatever entry point the data is being ingested into our database
+    if (positionLevel === 'CITY') {
+      positionLevel = 'LOCAL';
     }
-    if (races && typeof races.filter === 'function') {
-      const filtered = races.filter((option) => {
-        let positionLevel = option.position.level;
-        if (positionLevel === 'CITY') {
-          positionLevel = 'LOCAL';
-        }
+    const raceYear = (new Date(race.election.electionDay)).getFullYear()
 
-        if (value && (!updatedLevel || updatedLevel === '')) {
-          return option.position.name
-            .toLowerCase()
-            .includes(value.toLowerCase());
-        }
-        if (value && updatedLevel && updatedLevel !== '') {
-          return (
-            option.position.name.toLowerCase().includes(value.toLowerCase()) &&
-            positionLevel === updatedLevel.toUpperCase()
-          );
-        }
+    const textFiltered = inputValue ?
+      race.position.name
+      .toLowerCase()
+      .includes(inputValue.toLowerCase()) : true
 
-        return positionLevel === updatedLevel.toUpperCase();
-      });
-      setFilteredRaces(filtered);
-    }
-  };
+    const levelFiltered = level ?
+      level && positionLevel === level.toUpperCase() :true
+
+    const yearFiltered = yearFilter ?
+      parseInt(yearFilter) === raceYear : true
 
+    return textFiltered && levelFiltered && yearFiltered
+  }
+
+  const filtersEmpty = !level && !inputValue && !yearFilter
+
+  const filtered = filtersEmpty ?
+    races :
+    electionYears.reduce(
+      (racesGroupedByYear, year) => ({
+          ...racesGroupedByYear,
+          [year]: races[year]?.filter(filterRace)
+        }),
+      {}
+    )
```

Comment:
Let's not nest logic inside ternary operators.
Two separate lines with meaningful variable names.

Reply:
@tomer-tgp 

Not following what you mean by “Let’s not nest logic inside ternary operators” here? 🤔 

Thinking maybe you’re asking for the `.reduce()` to be stored into an earlier variable, but the entire point of using the ternary here would be to avoid running through the iteration needlessly, if indeed the filters are empty and we can just move forward w/ the collection of races that are already in memory…?

Or maybe you mean to have the reduce callback function abstracted out to a named function? But that reduction method isn’t used anywhere else so seems a bit redundant and maybe an extra layer of abstraction that’s not needed? 🤔 
I dunno.. just not sure what you want me to change here that would make sense..?

Reply:
Resolving w/ Slack convo here: https://goodpartyorg.slack.com/archives/C06NQMW4HJQ/p1711563062725549